Adds support for using Danger on PRs

orta opened the pull request:
**Summary**

Adds [Danger](https://github.com/danger/danger-js).

With initial Danger rules: 

* Warns on No new un-flowed JS file
* Fails build when FB Copyright header is not included

**Test plan**

This Dangerfile does not have the FB Copyright Header, so once Danger is set up she should fail this build.

**Discussion**

- [x] I based this on looking through a few PRs, will dig through a few more now that I've got something to show
- [x]  Danger relies on Jest's packages, should/can I make the deps local paths inside yarn? So there's no duplication confusion

**Setup**

This requires a bit of config by @cpojer - you will need to go through the "[creating a bot account](http://danger.systems/guides/getting_started.html#creating-a-bot-account-for-danger-to-use)"  - as this is OSS, it should be a new account with zero privileges to any repos. 

After that you'll need to add a token to travis for `DANGER_GITHUB_API_TOKEN`.

orta commented on dangerfile.js:
```diff
@@ -0,0 +1,48 @@
+// ("/href/thing", "name") to "<a href="/href/thing">name</a>"
+const createLink = (href: string, text: string): string => 
+  `<a href='${href}'>${text}</a>`;
+
```
some of these ^ are functions from Danger RB, in the future I should probably move them into DangerJS too - the use case seems pretty universal

cpojer replied:
yeah please do!

orta commented:
I think these CI fails are from our usage of `jest-transform` in Danger, so, this question becomes more important

> Danger relies on Jest's packages, should/can I make the deps local paths inside yarn? So there's no duplication confusion

orta commented on dangerfile.js:
```diff
@@ -0,0 +1,48 @@
+
+// ["1", "2", "3"] to "1, 2 and 3"
+const toSentence = (array: Array<string>) : string => 
+  array.slice(0, array.length - 1).join(', ') + ' and ' + array.pop();
```
oh yeah, definitely! Good catch.

cpojer commented on dangerfile.js:
```diff
@@ -0,0 +1,50 @@
+// New JS files should have the FB copyright header
+const noFBCopyrightFiles = newJsFiles.filter(filepath => {
+  const content = fs.readFileSync(filepath).toString();
+  return !content.includes('Facebook, Inc. All rights reserved');
```
could we make this more complete and check that the first thing in the file is the full FB copyright header?

orta replied:
Sure, might need to be a regex ( the year is variable ) this was the constant part, I can build on this

cpojer commented on package.json:
```diff
@@ -43,7 +45,7 @@
     "publish": "yarn run build-clean && yarn run build && lerna publish",
     "test": "yarn run typecheck && yarn run lint && yarn run build && yarn run jest && yarn run test-examples",
-    "test-ci": "yarn run typecheck && yarn run lint && yarn run build && yarn run jest-coverage -- -i && yarn run test-examples && node scripts/mapCoverage.js && codecov",
+    "test-ci": "yarn run typecheck && yarn run lint && yarn run build && yarn run jest-coverage -- -i && yarn run test-examples && node scripts/mapCoverage.js && codecov && yarn run danger",
```
I kinda wish danger could replace codecov.

orta replied:
I assume codecov reads some pre-existing files to derive output? It's totally feasible ( we have ruby Danger parsing JUnit files in our iOS app ) 

However, as you noted below, it's the state that doesn't necessarily work. In ruby Danger, there is a plugin for persistence (https://github.com/dbgrandi/danger-mcbrain) which could get replicated as a free heroku mongo somewhere or something to keep state

cpojer commented:
I'll get the access data for the jest-bot and will report back when I added the token.

A few things that would be nice (I will add more ideas some time):
* Fails if there are merge commits in the PR.
* Fails if there are changes to package.json without changes to yarn.lock.

Does Danger have any form of state? Can we potentially use it to compare the performance of the test run to master?

orta commented:
* Added a stricter rule check for the copyright header
* Added the package.json to yarn.lock check, but I made it `warn`, not `fail`, a version bump or changes to the `scripts` would trigger a fail.
* Turns out danger needs support for reading the commits, they're not in the raw diff, which makes sense considering you need an API for commit comments etc https://github.com/danger/danger-js/issues/81

cpojer commented:
@orta I added the token to travis using `jest-bot`. Is there anything else I need to do?

orta commented:
The [master CI is broken](https://travis-ci.org/facebook/jest/builds/189301422), I'm gonna test without running ESLinter et al

ArtsyOpenSource commented:
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>New JS files do not have the Facebook copyright header: <a href='https://github.com/orta/jest/blob/danger/dangerfile.js'>dangerfile.js</a></td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>Please ensure that @flow and 'use strict' are enabled on: <a href='https://github.com/orta/jest/blob/danger/dangerfile.js'>dangerfile.js</a></td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

orta commented on dangerfile.js:
```diff
@@ -0,0 +1,72 @@
+// @flow
+
+// const { danger, fail, warn } = require('danger');
```
Waiting on a release with https://github.com/danger/danger-js/pull/82 to support having this uncommented ( Danger normally comments out these requires and injects the deps in at runtime. It's only there for tooling support )

cpojer commented:
I set the "Display value in build log" to "on". Is there something else I need to do?

orta commented:
I mean at this point:

![screen shot 2017-01-05 at 5 37 35 pm](https://cloud.githubusercontent.com/assets/49038/21700279/ae238e26-d36d-11e6-8fff-29c5e481588b.png)

Everything else seems 👍

cpojer commented:
done!

orta commented:
Rock, that looks good to me. Will see if I can get enough time today to ship the commits API, otherwise might look over the weekend for that 👍

cpojer commented:
Awesome! Let me know when it's ready and doesn't spam people so much any more ;)

rogeliog commented:
@orta the Flow link warning has a link to `@flow` => https://github.com/flow.... should it be in backticks instead? or should it link to the Flow website?
